machineconfiguration/v1/types_machineosconfig.go

```diff
@@ -0,0 +1,243 @@
+	// +optional
+	ReleaseVersion string `json:"releaseVersion,omitempty"`
+	// containerFile describes the custom data the user has specified to build into the image.
+	// this is also commonly called a Dockerfile and you can treat it as such. The content is the content of your Dockerfile.
+	// you can specify up to 7 containerFiles
+	// +patchMergeKey=containerfileArch
+	// +patchStrategy=merge
+	// +listType=map
+	// +listMapKey=containerfileArch
+	// +kubebuilder:validation:MinItems=0
+	// +kubebuilder:validation:MaxItems=7
+	// +optional
+	Containerfile []MachineOSContainerfile `json:"containerFile" patchStrategy:"merge" patchMergeKey:"containerfileArch"`
+}
+
+// BuildOutputs holds all information needed to handle booting the image after a build
+// +union
+type BuildOutputs struct {
+	// currentImagePullSecret is the secret used to pull the final produced image.
+	// must live in the openshift-machine-config-operator namespace
+	// the final image push and pull secrets should be separate for security concerns. If the final image push secret is somehow exfiltrated,
+	// that gives someone the power to push images to the image repository. By comparison, if the final image pull secret gets exfiltrated,
+	// that only gives someone to pull images from the image repository. It's basically the principle of least permissions.
+	// this pull secret will be used on all nodes in the pool. These nodes will need to pull the final OS image and boot into it using rpm-ostree or bootc.
+	// +optional
+	CurrentImagePullSecret ImageSecretObjectReference `json:"currentImagePullSecret,omitempty"`
+}
+
+type MachineOSImageBuilder struct {
+	// imageBuilderType specifies the backend to be used to build the image.
+	// +kubebuilder:validation:Enum:=JobImageBuilder
+	// Valid options are: JobImageBuilder
+	ImageBuilderType MachineOSImageBuilderType `json:"imageBuilderType"`
+}
+
+// MachineOSContainerfile contains all custom content the user wants built into the image
+type MachineOSContainerfile struct {
+	// containerfileArch describes the architecture this containerfile is to be built for
+	// this arch is optional. If the user does not specify an architecture, it is assumed
+	// that the content can be applied to all architectures, or in a single arch cluster: the only architecture.
+	// +kubebuilder:validation:Enum:=arm64;amd64;ppc64le;s390x;aarch64;x86_64;noarch
+	// +kubebuilder:default:=noarch
+	// +optional
+	ContainerfileArch ContainerfileArch `json:"containerfileArch"`
+	// content is an embedded Containerfile/Dockerfile that defines the contents to be built into your image.
+	// for example, this would add the tree package to your hosts:
+	//   FROM configs AS final
+	//   RUN rpm-ostree install tree && \
+	//     ostree container commit
+	// +kubebuilder:validation:Required
+	Content string `json:"content"`
```

Comment:
We should add a length limit to this string, and avoid arbitrarily long content.

Is 4kb enough? 

```suggestion
	// +kubebuilder:validation:MaxLength=4096
	Content string `json:"content"`
```